### DISCUSSION TITLE
Allow only specified extra fields for models

### DISCUSSION THREAD

**Discussion by Cielquan** [pydantic/pydantic#3961] [Feature Request]
Currently you can only set the `extra` model config to allow or disallow (ignore/forbid are the _same_ in this case) additional fields. It would be nice if you could allow only a range of additional fields like so for example:

(The example assumes the default setting for `extra` -> `ignore`.)
```python
class Model(BaseModel, extra_fields=[("name", str)]):
    id: int

print(Model(id=1))                          # Output: id=1
print(Model(id=1, test=True))               # Output: id=1
print(Model(id=1, name="Hans"))             # Output: id=1 name='Hans'
print(Model(id=1, name="Hans", test=True))  # Output: id=1 name='Hans'
```

The field's name and type is a tuple and the `extra_fields` could contains multiple of these.

### Further behavior definition

As there is another config (`extra`) which changes behavior in this area the behavior must be defined if both are set:

#### `extra_fields=None` || `extra_fields=[]`
In this case the current behavior would apply.

#### `extra=Extra.allow`
If all extra fields are allowed the `extra_fields` setting can be ignored.

#### `extra=Extra.ignore`
Fields specified in the `extra_fields` list would be `Extra.allow` and anything else would be ignored like normal.

#### `extra=Extra.forbid`
Fields specified in the `extra_fields` list would be `Extra.allow` and anything else would trigger an Error like normal.

## Reason behind

Example Code (requires `pip install sqlmodel`):
```python
from typing import List, Optional

from pydantic import Extra
from sqlalchemy.orm import backref
from sqlmodel import Field, Relationship, SQLModel

class Node(SQLModel, table=True, extra=Extra.allow):
    id: int = Field(primary_key=True)
    parent_id: Optional[int] = Field(default=None, foreign_key='node.id')
    children: List['Node'] = Relationship(
        sa_relationship_kwargs=dict(
            backref=backref('parent', remote_side='Node.id'),
        ),
    )
```

I use [sqlmodel](https://sqlmodel.tiangolo.com/) which is a bridge between pydantic and sqlalchemy. I have a database model `Node` which has a self reference. The relationship between the `Node` models is one parent can have multiple children (one-to-many relationship). I can only get the relationship to work with the use of a `backref` which in turn dynamically creates the "backref-column" field, in this case called `parent`, based on the name (first arg to `backref()`) I give it. pydantic does of cause not know about this field and thus ignores it if set. Therefore I have to set `extra=Extra.allow` which in turn allows everything as an additional field which can be quite problematic. Because I know the added field's name (parent) beforehand it would be possible to restrict additional fields to this one.

**Comment by Chuangye-Wang** [comment]
One way is to set extra=allow and use @model_validator to check whether the extra input parameters in self.model_extra are the inputs you want. If some input parameters are not desired, raise errors.
